projects/feature-activation/0005-feature-activation-for-transactions.md

```diff
@@ -0,0 +1,196 @@
+
+## Retrieving Feature States for Transactions
+
+Analogously to the Feature Activation for Blocks, the state for a transaction will be retrieved from a method in the `FeatureService`. Considering the definition above, a reference-implementation is provided:
+
+```python
+class FeatureService:
+    def is_feature_active_for_transaction(self, *, tx: 'Transaction', feature: Feature) -> bool:
+        """Return whether a Feature is active for a certain Transaction."""
+        metadata = tx.get_metadata()
+        closest_block = self._tx_storage.get_block(not_none(metadata.closest_block))
+
+        return self.is_feature_active_for_block(block=closest_block, feature=feature)
+```
+
+## Handling reorgs
+
+Since the transaction's state is calculated based on an existing direct or indirect block dependency of the transaction, it's guaranteed that when this block is voided in a reorg, the transaction will also be voided via current consensus rules. Therefore, no specific handling of reorgs is necessary.
+
+It's also interesting to mention that since there's a requirement of 300 blocks between a transaction and the block it spends (via the Reward Lock mechanism), it's extremely unlikely that transactions are voided for this reason.
+
+## New `closest_block` metadata
+
+A new metadata field must be created to propagate the `closest_block` information forward in the DAG. Every time a transaction is received, its `closest_block` metadata must be calculated from its dependencies (parents and inputs). If the dependency is a block, the block itself is a `closest_block` candidate. If the dependency is a transaction, its own `closest_block` is a candidate. Given all candidates, the block with the greatest height is defined as the `closest_block`. This calculation is O(1) as the number of dependencies is bounded. It's guaranteed that this will be the same for all peers, as dependencies are part of the transaction's core structure and are validated. It's also required that the full node has all dependencies downloaded before accepting a new transaction.
+
+Here's a reference implementation for calculating `closest_block` metadata, defined in the `BaseTransaction` class:
+
+```python
+def _update_closest_block_metadata(self) -> None:
+    """
+    Set the tx's closest_block metadata.
+    For blocks, it's always None. For Transactions, it's the Block with the greatest height that is a direct
+    or indirect dependency (ancestor) of the transaction, including both funds and confirmation DAGs.
+    It's calculated by propagating the metadata forward in the DAG,
+    and it's used by Feature Activation for Transactions.
+    """
+    from hathor.transaction import Block, Transaction
+    if isinstance(self, Block):
+        return
+    assert isinstance(self, Transaction)
+    assert self.storage is not None
+    metadata = self.get_metadata()
+
+    if self.is_genesis:
+        metadata.closest_block = self._settings.GENESIS_BLOCK_HASH
```

Comment:
Should we point out this edge case in the description?

Reply:
Done!